### PR TITLE
CSP script nonce

### PR DESCRIPTION
**Pull request by bpeng**
## Proposed Changes

Resolves #GeoNet/www-geonet/issues/795

Changes proposed in this pull request:

- Add nonce to both script CSP and page handler to be used as script attribute

## Production Changes

The following production changes are required to deploy these changes:

- None

## Review

Check the box that applies to this code review.  If necessary please seek help with adding a checklist guide for the reviewer.
When assigning the code review please consider the expertise needed to review the changes.

- [ ] This is a content (documentation, web page etc) only change.
- [x] This is a minor change (meta data, bug fix, improve test coverage etc).
- [ ] This is a larger change (new feature, significant refactoring etc).  Please use the code review guidelines to add a checklist below to guide the code reviewer.

### Code Review Guide

*Insert check list here if needed.*

### CODE REVIEW

File: weft/handlers.go [hunk 1]
```diff
@@ -308,11 +323,14 @@ func setBestPracticeHeaders(w http.ResponseWriter, r *http.Request, customCsp ma
 				s = v1
 			}
 		}
-
 		csp.WriteString(k)
 		csp.WriteString(" ")
-		csp.WriteString(s)
-		csp.WriteString("; ")
+		if k == "script-src" && nonce != "" { //add nonce to CSP
+			csp.WriteString(fmt.Sprintf("'nonce-%s' 'strict-dynamic';", nonce))
```
Comment (CallumNZ) [inline weft/handlers.go:329]: Just checking if you need a space after the semicolon like on line 332.

Comment (bpeng) [reply weft/handlers.go:329]: no really necessary, but it leaves a space between different `src` items (last one is redundant)

**Comment by bpeng** [comment]
> Does our server check against nonce?
> If we're not validating it then why do we ask the browser to send it to us?

The `nonce` is generated randomly for each request, it is sent back to the page template to set the `nonce` attribute for the `<script>` elements, which need to be the same as in the CSP.

File: weft/handlers.go [hunk 2]
```diff
@@ -92,7 +97,11 @@ func MakeDirectHandler(rh DirectRequestHandler, eh ErrorHandler) http.HandlerFun
 // Responses are counted.  rh is not wrapped with a timer as this includes the write to the client.
 func MakeDirectHandlerWithCsp(rh DirectRequestHandler, eh ErrorHandler, customCsp map[string]string) http.HandlerFunc {
 	return func(w http.ResponseWriter, r *http.Request) {
-		n, err := rh(r, w)
+		nonce, err := getCspNonce(16)
+		if err != nil {
+			logger.Printf("setting error: %v", err)
+		}
+		n, err := rh(r, w, nonce)
```
[the remark this answers is not in the document]
Comment (bpeng) [reply weft/handlers.go:104]: hmm, this should go the error handler.

[the remark this answers is not in the document]
Comment (bpeng) [reply weft/handlers.go:104]: hmm, this should go the error handler.

**Comment by bpeng** [comment]
@junghao could you check [L96 ](https://github.com/GeoNet/kit/pull/74/files#diff-daee4c4bda54022c48c76f08f796b8daL96)is correct, when the **requestHandler** returns without error, it simply returns and skips setting Headers below, which is different from the other function at [L212](https://github.com/GeoNet/kit/pull/74/files#diff-daee4c4bda54022c48c76f08f796b8daL212).

The function **MakeDirectHandler** is only used in **FDSN**.

**Pull request merged**